File: ui/src/lib/zeigarnik.ts
/**
 * Client-side Zeigarnik effect proxy detector — port of crates/core/src/analysis/zeigarnik.rs.
 * Deterministic pattern matching only; no AI.
 */

import zeigarnikConfig from '../data/zeigarnik-config.json';
import type { CachedChapter } from './manuscriptCache';

export interface ZeigarnikConfig {
  cliffhanger_markers: string[];
  resolution_markers: string[];
  question_lead_ins: string[];
  short_fragment_max_words: number;
  min_gap_chapters_for_thread: number;
  max_total_mentions_for_thread: number;
  min_thread_term_len: number;
  top_threads_limit: number;
  min_question_words: number;
  max_questions_per_chapter: number;
}

export interface ZeigarnikChapterInput {
  path: string;
  title: string;
  content: string;
}

interface ChapterRow {
  chapter_index: number;
  file: string;
  title: string;
  word_count: number;
  sentence_count: number;
  question_count: number;
  ending_type: string;
  tension_score: number;
  ending_snippet: string;
}

interface ThreadRow {
  term: string;
  mention_count: number;
  first_chapter_index: number;
  first_file: string;
  first_snippet: string;
  gap_start_index: number;
  gap_end_index: number;
  max_gap_chapters: number;
  max_gap_words: number;
}

interface Occurrence {
  chapterIdx: number;
  sentenceInitial: boolean;
  snippet: string;
}

const STOPWORDS = new Set([
  'The', 'A', 'An', 'He', 'She', 'It', 'They', 'We', 'I', 'But', 'And', 'So',
  'Then', 'When', 'If', 'As', 'You', 'Chapter', 'There', 'Here', 'This', 'That', 'Her', 'His',
]);

function loadConfig(): ZeigarnikConfig {
  const t = zeigarnikConfig.thresholds;
  return {
    cliffhanger_markers: zeigarnikConfig.cliffhanger_markers,
    resolution_markers: zeigarnikConfig.resolution_markers,
    question_lead_ins: zeigarnikConfig.question_lead_ins,
    short_fragment_max_words: t.short_fragment_max_words,
    min_gap_chapters_for_thread: t.min_gap_chapters_for_thread,
    max_total_mentions_for_thread: t.max_total_mentions_for_thread,
    min_thread_term_len: t.min_thread_term_len,
    top_threads_limit: t.top_threads_limit,
    min_question_words: t.min_question_words,
    max_questions_per_chapter: t.max_questions_per_chapter,
  };
}

function stripMarkdown(raw: string): string {
  return raw
    .split('\n')
    .filter(l => !l.trimStart().startsWith('# '))
    .join('\n')
    .replace(/\*\*/g, '')
    .replace(/__/g, '');
}

function extractTitle(content: string): string | null {
  for (const line of content.split('\n').slice(0, 20)) {
    const t = line.trim();
    if (t.startsWith('#')) {
      return t.replace(/^#+\s*/, '');
    }
    if (t) return t;
  }
  return null;
}

function lastParagraph(body: string): string {
  const parts = body.split('\n\n').map(p => p.trim()).filter(Boolean);
  return parts.length ? parts[parts.length - 1] : '';
}

function splitSentences(body: string): string[] {
  const re = /[^.!?]*[.!?]+[\)\]\u201d'\u2019\u201c\u0022]*/gu;
  const out: string[] = [];
  let matchedLen = 0;
  for (const m of body.matchAll(re)) {
    const s = m[0].trim();
    if (s) {
      out.push(s);
      matchedLen += s.length;
    }
  }
  const remainder = body.trim();
  if (remainder.length > matchedLen + 20) {
    const tail = remainder.split(/[.!?]/).pop()?.trim() ?? '';
    if (tail.split(/\s+/).length >= 3) {
      out.push(tail);
    }
  }
  return out;
}

function detectEnding(
  lastPara: string,
  sentences: string[],
  config: ZeigarnikConfig,
): { endingType: string; tensionScore: number; endingSnippet: string } {
  const lower = lastPara.toLowerCase();
  let score = 35;

  const lastSentence = sentences.length ? sentences[sentences.length - 1] : '';
  const lastLower = lastSentence.toLowerCase();

  if (lastLower.trimEnd().endsWith('?')) score += 25;
  if (lastPara.trimEnd().endsWith('\u2014') || lastPara.trimEnd().endsWith('...') || lastPara.trimEnd().endsWith('\u2026')) {
    score += 15;
  }
  if (config.cliffhanger_markers.some(m => lower.includes(m))) score += 25;
  if (config.resolution_markers.some(m => lower.includes(m))) score -= 35;

  if (sentences.length >= 2) {
    const lastWords = lastSentence.split(/\s+/).length;
    const priorWords = sentences[sentences.length - 2].split(/\s+/).length;
    if (lastWords > 0 && lastWords <= config.short_fragment_max_words && priorWords > lastWords * 2) {
      score += 12;
    }
  }

  score = Math.max(0, Math.min(100, score));
  const endingType = score >= 65 ? 'cliffhanger' : score <= 25 ? 'resolved' : 'neutral';

  const take = sentences.length >= 2 ? 2 : Math.min(sentences.length, 1);
  let snippet = sentences.slice(Math.max(0, sentences.length - take)).join(' ');
  if (!snippet) snippet = [...lastPara].slice(0, 200).join('');
  if ([...snippet].length > 240) {
    snippet = [...snippet].slice(0, 240).join('') + '…';
  }

  return { endingType, tensionScore: score, endingSnippet: snippet };
}

function extractOpenQuestions(sentences: string[], config: ZeigarnikConfig): string[] {
  const candidates: { leadIn: boolean; text: string }[] = [];
  for (const s of sentences) {
    if (!s.trimEnd().endsWith('?')) continue;
    const wordCount = s.split(/\s+/).length;
    if (wordCount < config.min_question_words) continue;
    const lower = s.toLowerCase();
    const isLeadIn = config.question_lead_ins.some(p => lower.includes(p));
    candidates.push({ leadIn: isLeadIn, text: s.trim() });
  }
  candidates.sort((a, b) => Number(b.leadIn) - Number(a.leadIn));
  return candidates.slice(0, config.max_questions_per_chapter).map(c => c.text);
}

function scanEntityOccurrences(chapterTexts: string[]): Map<string, Occurrence[]> {
  const capWord = String.raw`[A-Z][a-zA-Z'\u2019]+`;
  const re = new RegExp(String.raw`(?:${capWord})(?:\s+(?:${capWord})){0,2}`, 'g');
  const map = new Map<string, Occurrence[]>();

  chapterTexts.forEach((text, chapterIdx) => {
    for (const m of text.matchAll(re)) {
      const raw = m[0];
      const words = raw.split(/\s+/);
      const isMultiword = words.length > 1;
      if (!isMultiword && STOPWORDS.has(raw)) continue;

      const start = m.index ?? 0;
      const before = text.slice(0, start).trimEnd();
      const lastChar = before.length ? before[before.length - 1] : null;
      const sentenceInitial = lastChar === null
        || lastChar === '.' || lastChar === '!' || lastChar === '?'
        || lastChar === '\u201c' || lastChar === '\u2018' || lastChar === '"';

      const norm = raw.toLowerCase();
      const ctxStart = Math.max(0, start - 40);
      const ctxEnd = Math.min(text.length, start + raw.length + 40);
      const snippet = `…${text.slice(ctxStart, ctxEnd).trim()}…`;

      const list = map.get(norm) ?? [];
      list.push({ chapterIdx, sentenceInitial, snippet });
      map.set(norm, list);
    }
  });

  return map;
}

function titleCase(s: string): string {
  return s.split(/\s+/).map(w => {
    if (!w) return '';
    return w[0].toUpperCase() + w.slice(1);
  }).join(' ');
}

function findOpenThreads(
  occurrences: Map<string, Occurrence[]>,
  chapterRows: ChapterRow[],
  config: ZeigarnikConfig,
): ThreadRow[] {
  const threads: ThreadRow[] = [];

  for (const [term, occs] of occurrences) {
    if (term.length < config.min_thread_term_len) continue;

    const isMultiword = term.includes(' ');
    const hasNonInitial = occs.some(o => !o.sentenceInitial);
    if (!isMultiword && !hasNonInitial) continue;

    let chapterIdxs = [...new Set(occs.map(o => o.chapterIdx))].sort((a, b) => a - b);
    if (chapterIdxs.length < 2) continue;
    if (chapterIdxs.length > config.max_total_mentions_for_thread) continue;

    let maxGap = 0;
    let gapStart = 0;
    let gapEnd = 0;
    for (let i = 0; i < chapterIdxs.length - 1; i++) {
      const g = chapterIdxs[i + 1] - chapterIdxs[i];
      if (g > maxGap) {
        maxGap = g;
        gapStart = chapterIdxs[i];
        gapEnd = chapterIdxs[i + 1];
      }
    }
    if (maxGap < config.min_gap_chapters_for_thread) continue;

    const gapWords = chapterRows
      .filter(c => c.chapter_index > gapStart && c.chapter_index < gapEnd)
      .reduce((sum, c) => sum + c.word_count, 0);

    const firstOcc = occs.find(o => o.chapterIdx === chapterIdxs[0]);
    const firstFile = chapterRows[chapterIdxs[0]]?.file ?? '';

    threads.push({
      term: titleCase(term),
      mention_count: chapterIdxs.length,
      first_chapter_index: chapterIdxs[0],
      first_file: firstFile,
      first_snippet: firstOcc?.snippet ?? '',
      gap_start_index: gapStart,
      gap_end_index: gapEnd,
      max_gap_chapters: maxGap,
      max_gap_words: gapWords,
    });
  }

  threads.sort((a, b) =>
    b.max_gap_chapters - a.max_gap_chapters || b.max_gap_words - a.max_gap_words,
  );
  return threads.slice(0, config.top_threads_limit);
}

export function cachedChapterToInput(c: CachedChapter): ZeigarnikChapterInput {
  return {
    path: c.path,
    title: extractTitle(c.content) ?? c.path,
    content: c.content,
  };
}

/** Run Zeigarnik analysis on ordered chapter inputs. Returns zeigarnik_v1 JSON object. */
export function runZeigarnikAnalysis(chapters: ZeigarnikChapterInput[]): Record<string, unknown> {
  if (chapters.length === 0) {
    throw new Error('No chapters to analyze.');
  }

  const config = loadConfig();
  const chapterTexts: string[] = [];
  const chapterRows: ChapterRow[] = [];

  chapters.forEach((chapter, i) => {
    const fname = chapter.path || `chapter-${i + 1}`;
    const title = chapter.title || extractTitle(chapter.content) || fname;
    const body = stripMarkdown(chapter.content);
    const wordCount = body.split(/\s+/).filter(Boolean).length;
    const sentences = splitSentences(body);
    const questions = extractOpenQuestions(sentences, config);
    const lastPara = lastParagraph(body);
    const { endingType, tensionScore, endingSnippet } = detectEnding(lastPara, sentences, config);

    chapterRows.push({
      chapter_index: i,
      file: fname,
      title,
      word_count: wordCount,
      sentence_count: sentences.length,
      question_count: questions.length,
      ending_type: endingType,
      tension_score: tensionScore,
      ending_snippet: endingSnippet,
    });
    chapterTexts.push(body);
  });

  const occurrences = scanEntityOccurrences(chapterTexts);
  const threads = findOpenThreads(occurrences, chapterRows, config);

  const totalChapters = chapterRows.length;
  const totalWords = chapterRows.reduce((s, c) => s + c.word_count, 0);
  const cliffhangerCount = chapterRows.filter(c => c.ending_type === 'cliffhanger').length;
  const resolvedCount = chapterRows.filter(c => c.ending_type === 'resolved').length;
  const totalQuestions = chapterRows.reduce((s, c) => s + c.question_count, 0);
  const avgTension = totalChapters > 0
    ? chapterRows.reduce((s, c) => s + c.tension_score, 0) / totalChapters
    : 0;
  const longestGapChapters = threads.reduce((m, t) => Math.max(m, t.max_gap_chapters), 0);

  const cliffhangerPct = totalChapters > 0 ? (cliffhangerCount / totalChapters) * 100 : 0;
  const chaptersWithQuestions = chapterRows.filter(c => c.question_count > 0).length;
  const questionPct = totalChapters > 0 ? (chaptersWithQuestions / totalChapters) * 100 : 0;
  const threadPct = threads.length === 0
    ? 0
    : Math.min(100, (Math.min(threads.length, totalChapters) / totalChapters) * 100);
  const overallZeigarnikPct = Math.round((cliffhangerPct + questionPct + threadPct) / 3);

  return {
    schema: 'zeigarnik_v1',
    note: 'Textual proxy analysis only — this measures manuscript structure (unresolved endings, open questions, long-gap threads), not reader recall itself. No AI was used to generate these results.',
    summary: {
      overall_zeigarnik_pct: overallZeigarnikPct,
      total_chapters: totalChapters,
      total_words: totalWords,
      cliffhanger_endings: cliffhangerCount,
      resolved_endings: resolvedCount,
      cliffhanger_pct: Math.round(cliffhangerPct),
      total_open_questions: totalQuestions,
      avg_tension_score: Math.round(avgTension * 10) / 10,
      open_thread_count: threads.length,
      longest_gap_chapters: longestGapChapters,
    },
    chapters: chapterRows,
    threads,
  };
}

File: src-tauri/src/analysis/zeigarnik.rs
// analysis/zeigarnik.rs — Zeigarnik-effect proxy detector for the Craft platform.
//
// This is a deliberately AI-free, deterministic text scanner. The Zeigarnik
// effect itself — people recall interrupted tasks better than completed ones
// — is a claim about what happens in a reader's mind, which no algorithm can
// observe directly. What this module measures instead are textual proxies
// craft writers use to court that effect:
//
//   1. Chapter-ending tension  — does a chapter end on an unresolved beat
//      (a question, a cliffhanger phrase, an abrupt short fragment) or a
//      resolved one?
//   2. Open narrative questions — sentences that raise a question the prose
//      doesn't answer in the same breath.
//   3. Long-gap threads — a capitalized term/phrase (proxy for a named
//      person, place, or object) that appears, then goes quiet for several
//      chapters, then resurfaces. That gap is the shape of an open loop.
//
// None of this proves a reader will actually experience heightened recall —
// it surfaces where the manuscript's structure matches the pattern, so the
// author can judge for themselves. All phrase lists and thresholds are
// loaded from the zeigarnik_config table (see db.rs) rather than hardcoded.

use std::collections::HashMap;
use std::path::PathBuf;
use tauri::{AppHandle, Manager};
use regex::Regex;

use super::{emit, err, GenreResult};
use super::chapters::{collect_chapters, extract_title};
use crate::db;

// ── Request / command ───────────────────────────────────────────────────────

#[derive(serde::Deserialize)]
pub struct ZeigarnikRequest {
    pub folder: String,
}

#[tauri::command]
pub async fn analyze_zeigarnik_for_story(app: AppHandle, request: ZeigarnikRequest) -> GenreResult {
    let folder = PathBuf::from(&request.folder);
    if !folder.exists() { return err("Folder does not exist."); }

    crate::reset_cancel();
    let chapter_paths = collect_chapters(&folder);
    if chapter_paths.is_empty() { return err("No .md files found."); }

    let database = app.state::<db::Db>();
    let config = { let conn = database.0.lock().unwrap(); db::load_zeigarnik_config(&conn) };
    let run_ts = chrono::Utc::now().to_rfc3339();

    emit(&app, &format!("Found {} chapter file(s). Scanning for open loops (no AI — pattern matching only)...", chapter_paths.len()));

    // ── Pass 1: read + per-chapter metrics + raw text for entity scan ──────
    let mut chapter_texts: Vec<String> = Vec::with_capacity(chapter_paths.len());
    let mut chapter_rows: Vec<db::ZeigarnikChapterRow> = Vec::with_capacity(chapter_paths.len());

    for (i, path) in chapter_paths.iter().enumerate() {
        let fname = path.file_name().unwrap_or_default().to_string_lossy().to_string();
        let raw = match std::fs::read_to_string(path) {
            Ok(c) => c,
            Err(e) => { emit(&app, &format!("  ⚠ Could not read {}: {}", fname, e)); String::new() }
        };
        let title = extract_title(&raw).unwrap_or_else(|| fname.clone());
        let body = strip_markdown(&raw);
        let word_count = body.split_whitespace().count();
        let sentences = split_sentences(&body);
        let questions = extract_open_questions(&sentences, &config);
        let last_para = last_paragraph(&body);
        let (ending_type, tension_score, ending_snippet) = detect_ending(&last_para, &sentences, &config);

        emit(&app, &format!("  [{}/{}] {} — {} words, {} question(s), {} ending",
            i + 1, chapter_paths.len(), fname, word_count, questions.len(), ending_type));

        chapter_rows.push(db::ZeigarnikChapterRow {
            chapter_index: i as i64,
            file: fname,
            title,
            word_count: word_count as i64,
            sentence_count: sentences.len() as i64,
            question_count: questions.len() as i64,
            ending_type,
            tension_score: tension_score as i64,
            ending_snippet,
        });
        chapter_texts.push(body);

        if crate::is_cancelled() { emit(&app, "⚠ Cancelled."); return err("Cancelled."); }
    }

    // ── Pass 2: track candidate open-loop terms across chapters ────────────
    emit(&app, "Tracking recurring names, places, and objects across chapters...");
    let occurrences = scan_entity_occurrences(&chapter_texts);
    let threads = find_open_threads(&occurrences, &chapter_rows, &config);
    emit(&app, &format!("  {} candidate open thread(s) found (gap ≥ {} chapters).", threads.len(), config.min_gap_chapters_for_thread));

    // ── Persist ──────────────────────────────────────────────────────────
    {
        let conn = database.0.lock().unwrap();
        if let Err(e) = db::replace_zeigarnik_analysis(&conn, &request.folder, &chapter_rows, &threads) {
            return err(&format!("Could not save analysis: {}", e));
        }
    }

    // ── Summary + JSON document ─────────────────────────────────────────
    let total_chapters = chapter_rows.len();
    let total_words: i64 = chapter_rows.iter().map(|c| c.word_count).sum();
    let cliffhanger_count = chapter_rows.iter().filter(|c| c.ending_type == "cliffhanger").count();
    let resolved_count = chapter_rows.iter().filter(|c| c.ending_type == "resolved").count();
    let total_questions: i64 = chapter_rows.iter().map(|c| c.question_count).sum();
    let avg_tension = if total_chapters > 0 {
        chapter_rows.iter().map(|c| c.tension_score).sum::<i64>() as f64 / total_chapters as f64
    } else { 0.0 };
    let longest_gap_chapters = threads.iter().map(|t| t.max_gap_chapters).max().unwrap_or(0);

    // ── Overall Zeigarnik usage percentage ──────────────────────────────
    // Composite of three proxies, equally weighted:
    //   1. % of chapters ending on a cliffhanger (vs resolved)
    //   2. % of chapters that contain at least one open question
    //   3. Whether open threads exist (binary scaled: thread_count > 0 gives a boost)
    let cliffhanger_pct = if total_chapters > 0 { cliffhanger_count as f64 / total_chapters as f64 * 100.0 } else { 0.0 };
    let chapters_with_questions = chapter_rows.iter().filter(|c| c.question_count > 0).count();
    let question_pct = if total_chapters > 0 { chapters_with_questions as f64 / total_chapters as f64 * 100.0 } else { 0.0 };
    let thread_pct = if threads.is_empty() { 0.0 } else { (threads.len().min(total_chapters) as f64 / total_chapters as f64 * 100.0).min(100.0) };
    let overall_zeigarnik_pct = ((cliffhanger_pct + question_pct + thread_pct) / 3.0).round();

    let json = serde_json::json!({
        "schema": "zeigarnik_v1",
        "note": "Textual proxy analysis only — this measures manuscript structure (unresolved endings, open questions, long-gap threads), not reader recall itself. No AI was used to generate these results.",
        "summary": {
            "overall_zeigarnik_pct": overall_zeigarnik_pct,
            "total_chapters": total_chapters,
            "total_words": total_words,
            "cliffhanger_endings": cliffhanger_count,
            "resolved_endings": resolved_count,
            "cliffhanger_pct": cliffhanger_pct.round(),
            "total_open_questions": total_questions,
            "avg_tension_score": (avg_tension * 10.0).round() / 10.0,
            "open_thread_count": threads.len(),
            "longest_gap_chapters": longest_gap_chapters,
        },
        "chapters": chapter_rows.iter().map(|c| serde_json::json!({
            "chapter_index": c.chapter_index,
            "file": c.file,
            "title": c.title,
            "word_count": c.word_count,
            "sentence_count": c.sentence_count,
            "question_count": c.question_count,
            "ending_type": c.ending_type,
            "tension_score": c.tension_score,
            "ending_snippet": c.ending_snippet,
        })).collect::<Vec<_>>(),
        "threads": threads.iter().map(|t| serde_json::json!({
            "term": t.term,
            "mention_count": t.mention_count,
            "first_chapter_index": t.first_chapter_index,
            "first_file": t.first_file,
            "first_snippet": t.first_snippet,
            "gap_start_index": t.gap_start_index,
            "gap_end_index": t.gap_end_index,
            "max_gap_chapters": t.max_gap_chapters,
            "max_gap_words": t.max_gap_words,
        })).collect::<Vec<_>>(),
    });
    let content = json.to_string();

    { let conn = database.0.lock().unwrap(); let _ = db::save_document_at(&conn, &request.folder, "zeigarnik_analysis", &content, &run_ts); }
    emit(&app, "✓ Zeigarnik analysis saved to database.");

    GenreResult { success: true, report: content, error: String::new(), run_ts }
}

// ── Text prep ────────────────────────────────────────────────────────────────

/// Strip the leading "# Title" line and light markdown formatting so pattern
/// matching runs against plain prose.
fn strip_markdown(raw: &str) -> String {
    raw.lines()
        .filter(|l| !l.trim_start().starts_with("# "))
        .collect::<Vec<_>>()
        .join("\n")
        .replace("**", "")
        .replace("__", "")
}

fn last_paragraph(body: &str) -> String {
    body.split("\n\n")
        .map(|p| p.trim())
        .filter(|p| !p.is_empty())
        .last()
        .unwrap_or("")
        .to_string()
}

/// Naive sentence splitter: good enough for pattern scanning, not linguistics.
/// Splits on ./!/? runs followed by whitespace, keeping the terminator.
fn split_sentences(body: &str) -> Vec<String> {
    let re = Regex::new(r"[^.!?]*[.!?]+[\)\]\u{201d}'\u{2019}\u{201c}\u{22}]*").unwrap();
    let mut out: Vec<String> = re.find_iter(body)
        .map(|m| m.as_str().trim().to_string())
        .filter(|s| !s.is_empty())
        .collect();

    // Catch a trailing fragment with no terminal punctuation (common at the
    // very end of a chapter file).
    let matched_len: usize = out.iter().map(|s| s.len()).sum::<usize>();
    let remainder = body.trim();
    if remainder.len() > matched_len + 20 {
        if let Some(tail) = remainder.rsplit(['.', '!', '?']).next() {
            let tail = tail.trim();
            if tail.split_whitespace().count() >= 3 {
                out.push(tail.to_string());
            }
        }
    }
    out
}

// ── Chapter-ending tension scoring ──────────────────────────────────────────

fn detect_ending(last_para: &str, sentences: &[String], config: &db::ZeigarnikConfig) -> (String, i64, String) {
    let lower = last_para.to_lowercase();
    let mut score: i64 = 35; // neutral baseline

    let last_sentence = sentences.last().cloned().unwrap_or_default();
    let last_lower = last_sentence.to_lowercase();

    if last_lower.trim_end().ends_with('?') {
        score += 25;
    }
    if last_para.trim_end().ends_with('\u{2014}') || last_para.trim_end().ends_with("...") || last_para.trim_end().ends_with('\u{2026}') {
        score += 15;
    }
    if config.cliffhanger_markers.iter().any(|m| lower.contains(m.as_str())) {
        score += 25;
    }
    if config.resolution_markers.iter().any(|m| lower.contains(m.as_str())) {
        score -= 35;
    }
    if sentences.len() >= 2 {
        let last_words = last_sentence.split_whitespace().count();
        let prior_words = sentences[sentences.len() - 2].split_whitespace().count();
        if last_words > 0 && last_words <= config.short_fragment_max_words && prior_words > last_words * 2 {
            score += 12;
        }
    }

    let score = score.clamp(0, 100);
    let ending_type = if score >= 65 { "cliffhanger" } else if score <= 25 { "resolved" } else { "neutral" };

    // Snippet: last one or two sentences, capped.
    let take = if sentences.len() >= 2 { 2 } else { sentences.len().min(1) };
    let mut snippet = sentences[sentences.len().saturating_sub(take)..].join(" ");
    if snippet.is_empty() { snippet = last_para.chars().take(200).collect(); }
    if snippet.chars().count() > 240 {
        snippet = snippet.chars().take(240).collect::<String>() + "…";
    }

    (ending_type.to_string(), score, snippet)
}

// ── Open narrative questions ────────────────────────────────────────────────

fn extract_open_questions(sentences: &[String], config: &db::ZeigarnikConfig) -> Vec<String> {
    // Collect every question sentence long enough to be more than a dialogue
    // tag, then prefer ones that read like a genuine open narrative question
    // ("What if...", "Why would...") over throwaway quips ("You okay?") when
    // trimming down to the per-chapter cap. Order within each group is
    // preserved (stable sort) so earlier questions still come first.
    let mut candidates: Vec<(bool, String)> = Vec::new();
    for s in sentences {
        if !s.trim_end().ends_with('?') { continue; }
        let word_count = s.split_whitespace().count();
        if word_count < config.min_question_words { continue; }
        let lower = s.to_lowercase();
        let is_lead_in = config.question_lead_ins.iter().any(|p| lower.contains(p.as_str()));
        candidates.push((is_lead_in, s.trim().to_string()));
    }
    candidates.sort_by(|a, b| b.0.cmp(&a.0));
    candidates.into_iter().take(config.max_questions_per_chapter).map(|(_, s)| s).collect()
}

// ── Entity / recurring-term tracking (for long-gap open threads) ───────────

struct Occurrence {
    chapter_idx: usize,
    sentence_initial: bool,
    snippet: String,
}

/// Find 1–3-word capitalized sequences (proxy for proper nouns / named
/// objects) in each chapter, noting whether each occurrence sits at the
/// start of a sentence (where capitalization is just grammar, not a name).
fn scan_entity_occurrences(chapter_texts: &[String]) -> HashMap<String, Vec<Occurrence>> {
    let cap_word = r"[A-Z][a-zA-Z'\u{2019}]+";
    let re = Regex::new(&format!(r"(?:{cw})(?:\s+(?:{cw})){{0,2}}", cw = cap_word)).unwrap();
    let stopwords: [&str; 24] = [
        "The", "A", "An", "He", "She", "It", "They", "We", "I", "But", "And", "So",
        "Then", "When", "If", "As", "You", "Chapter", "There", "Here", "This", "That", "Her", "His",
    ];

    let mut map: HashMap<String, Vec<Occurrence>> = HashMap::new();

    for (chapter_idx, text) in chapter_texts.iter().enumerate() {
        for m in re.find_iter(text) {
            let raw = m.as_str();
            let words: Vec<&str> = raw.split_whitespace().collect();
            let is_multiword = words.len() > 1;

            if !is_multiword && stopwords.contains(&raw) { continue; }

            // Sentence-initial if the preceding non-space char is a sentence
            // terminator, a quote, or we're at the start of the text.
            let start = m.start();
            let sentence_initial = {
                let before = text[..start].trim_end();
                match before.chars().last() {
                    None => true,
                    Some(c) => matches!(c, '.' | '!' | '?' | '\u{201c}' | '\u{2018}' | '"'),
                }
            };

            let norm = raw.to_lowercase();
            let ctx_start = floor_char_boundary(text, start.saturating_sub(40));
            let ctx_end = ceil_char_boundary(text, (m.end() + 40).min(text.len()));
            let snippet = format!("…{}…", text[ctx_start..ctx_end].trim());

            map.entry(norm).or_default().push(Occurrence { chapter_idx, sentence_initial, snippet });
        }
    }

    map
}

fn find_open_threads(
    occurrences: &HashMap<String, Vec<Occurrence>>,
    chapter_rows: &[db::ZeigarnikChapterRow],
    config: &db::ZeigarnikConfig,
) -> Vec<db::ZeigarnikThreadRow> {
    let mut threads = Vec::new();

    for (term, occs) in occurrences {
        if term.len() < config.min_thread_term_len { continue; }

        let is_multiword = term.contains(' ');
        let has_non_initial = occs.iter().any(|o| !o.sentence_initial);
        // Single capitalized words that only ever appear at a sentence start
        // are almost always just grammar (start-of-sentence capitalization),
        // not a genuine recurring name/object — skip them.
        if !is_multiword && !has_non_initial { continue; }

        let mut chapter_idxs: Vec<usize> = occs.iter().map(|o| o.chapter_idx).collect();
        chapter_idxs.sort_unstable();
        chapter_idxs.dedup();

        if chapter_idxs.len() < 2 { continue; }
        if chapter_idxs.len() > config.max_total_mentions_for_thread { continue; } // ubiquitous — a main character, not an open loop

        let mut max_gap = 0usize;
        let mut gap_start = 0usize;
        let mut gap_end = 0usize;
        for w in chapter_idxs.windows(2) {
            let g = w[1] - w[0];
            if g > max_gap { max_gap = g; gap_start = w[0]; gap_end = w[1]; }
        }
        if max_gap < config.min_gap_chapters_for_thread { continue; }

        let gap_words: i64 = chapter_rows.iter()
            .filter(|c| (c.chapter_index as usize) > gap_start && (c.chapter_index as usize) < gap_end)
            .map(|c| c.word_count)
            .sum();

        let first_occ = occs.iter().find(|o| o.chapter_idx == chapter_idxs[0]);
        let first_file = chapter_rows.get(chapter_idxs[0]).map(|c| c.file.clone()).unwrap_or_default();

        // Render with original casing from the first occurrence's snippet where possible.
        let display_term = title_case(term);

        threads.push(db::ZeigarnikThreadRow {
            term: display_term,
            mention_count: chapter_idxs.len() as i64,
            first_chapter_index: chapter_idxs[0] as i64,
            first_file,
            first_snippet: first_occ.map(|o| o.snippet.clone()).unwrap_or_default(),
            gap_start_index: gap_start as i64,
            gap_end_index: gap_end as i64,
            max_gap_chapters: max_gap as i64,
            max_gap_words: gap_words,
        });
    }

    threads.sort_by(|a, b| {
        b.max_gap_chapters.cmp(&a.max_gap_chapters)
            .then(b.max_gap_words.cmp(&a.max_gap_words))
    });
    threads.truncate(config.top_threads_limit);
    threads
}

fn floor_char_boundary(s: &str, mut i: usize) -> usize {
    while i > 0 && !s.is_char_boundary(i) { i -= 1; }
    i
}

fn ceil_char_boundary(s: &str, mut i: usize) -> usize {
    while i < s.len() && !s.is_char_boundary(i) { i += 1; }
    i
}

fn title_case(s: &str) -> String {
    s.split_whitespace()
        .map(|w| {
            let mut c = w.chars();
            match c.next() {
                Some(first) => first.to_uppercase().collect::<String>() + c.as_str(),
                None => String::new(),
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}
